### PR TITLE
[Wallet] Analytics Retooling + Analytics for Send & Request

### PR DESCRIPTION
**Pull request by tarikbellamine**
### Description
Created ability to add event properties for subEvent tracking and exposed those methods to the component used in the wallet.

Added analytics for the 2nd half of the Send flow (Ivan already did the first half), Secure Send, and the Request Payment flow.

### Other changes
Minor cleanup of dead code.

### Tested
No new tests added.

### Related issues
#3645 & #4008 

### Backwards compatibility
Yes

### CODE REVIEW

File: packages/mobile/src/navigator/Navigator.tsx [hunk 1]
```diff
@@ -229,18 +233,37 @@ const sendAmountScreenOptions = ({
...
+    headerTitle: () => <HeaderTitleWithBalance title={title} token={CURRENCY_ENUM.DOLLAR} />,
   }
 }
 
+const validateRecipientIntroScreenOptions = () => ({
+  ...emptyHeader,
+  headerLeft: () => <CancelButton eventName={CustomEventNames.send_secure_cancel} />,
+})
+
+const validateRecipientAccountScreenOptions = () => ({
```
Comment (aaronmgdr) [inline packages/mobile/src/navigator/Navigator.tsx]: nit: if these are only used by one screen it might be better to define them in its file as this file will become huge if every screen has a ScreenOptions defined here

File: packages/mobile/src/send/SendConfirmation.tsx [hunk 2]
```diff
@@ -148,6 +157,18 @@ export class SendConfirmation extends React.Component<Props, State> {
 
     const timestamp = Date.now()
 
+    CeloAnalytics.track(CustomEventNames.send_confirm, {
+      method: this.props.route.params?.isFromScan ? 'scan' : 'search',
```
Comment (aaronmgdr) [inline packages/mobile/src/send/SendConfirmation.tsx:168]: does this win for the most custom properties tracked?

File: packages/react-components/analytics/CeloAnalytics.tsx [hunk 3]
```diff
@@ -116,53 +121,63 @@ class CeloAnalytics {
...
-      new Map<string, number>([['__startTrackingTime__', Date.now()]])
+      new Map<string, SubEventData>([
+        ['__startTracking__', { timestamp: Date.now(), subEventProps: eventProperties }],
+      ])
     )
   }
 
   // See startTracking
-  trackSubEvent(eventName: string, subEventName: string) {
+  trackSubEvent(eventName: string, subEventName: string, eventProperties = {}) {
```
Comment (aaronmgdr) [inline packages/react-components/analytics/CeloAnalytics.tsx:141]: should eventProperties be typed here to only allow certain keys?

File: packages/mobile/src/navigator/Navigator.tsx (hunk 1 above)
Comment (tarikbellamine) [reply packages/mobile/src/navigator/Navigator.tsx]: I think the issue is that there is no easy way to pass it into the `Navigator` file. Seems like defining them here is what we've been doing for now but can definitely see a point where it might make sense to break them out into their own file

File: packages/mobile/src/send/SendConfirmation.tsx (hunk 2 above)
Comment (tarikbellamine) [reply packages/mobile/src/send/SendConfirmation.tsx:168]: Can you elaborate?

Comment (aaronmgdr) [reply packages/mobile/src/send/SendConfirmation.tsx:168]: just the number of things being tracked is impressive.

File: packages/react-components/analytics/CeloAnalytics.tsx (hunk 3 above)
Comment (tarikbellamine) [reply packages/react-components/analytics/CeloAnalytics.tsx:141]: Yeah that's a good suggestion. Will do

File: packages/mobile/src/send/SendConfirmation.tsx (hunk 2 above)
Comment (tarikbellamine) [reply packages/mobile/src/send/SendConfirmation.tsx:168]: Yeah, we've certainly turned the analytics up a notch!

File: packages/mobile/src/navigator/Navigator.tsx (hunk 1 above)
Comment (aaronmgdr) [reply packages/mobile/src/navigator/Navigator.tsx]: I dont follow that there is no easy way to pass it into the Navigator.

Comment (tarikbellamine) [reply packages/mobile/src/navigator/Navigator.tsx]: I think I misunderstood. Do you mean just define the options on the component screen  but just export it to `Navigator`?

Comment (aaronmgdr) [reply packages/mobile/src/navigator/Navigator.tsx]: like this 
https://github.com/celo-org/celo-monorepo/blob/master/packages/mobile/src/backup/BackupPhrase.tsx#L49

File: packages/mobile/src/backup/BackupPhrase.tsx [hunk 4]
```diff
@@ -1,3 +1,4 @@
+import { CustomEventNames } from '@celo/react-components/analytics/constants'
```
Comment (aaronmgdr) [inline packages/mobile/src/backup/BackupPhrase.tsx]: why this change?

Comment (tarikbellamine) [reply packages/mobile/src/backup/BackupPhrase.tsx]: Was going to consolidate events into `react-components` but have since decided against it

File: packages/mobile/src/account/Education.tsx [hunk 5]
```diff
@@ -84,7 +84,7 @@ class Education extends React.Component<Props, State> {
 
     return (
       <View style={style.body}>
-        {/*
```
Comment (aaronmgdr) [inline packages/mobile/src/account/Education.tsx:87]: lol

File: packages/mobile/src/analytics/constants.ts [hunk 6]
```diff
@@ -239,68 +251,76 @@ export enum CommonValues {
...
-  'testnet',
-  'timeElapsed',
-  'title',
-  'to',
-  'tti',
-  'txId',
-  'verificationIndex',
-  'verificationsRemaining',
-]
+export enum PropertyPathWhitelist {
```
Comment (aaronmgdr) [inline packages/mobile/src/analytics/constants.ts:263]: oh nice

File: packages/mobile/src/paymentRequest/PaymentRequestUnavailable.tsx [hunk 7]
```diff
@@ -1,16 +1,28 @@
...
 
 type RouteProps = StackScreenProps<StackParamList, Screens.PaymentRequestUnavailable>
 type Props = RouteProps
 
+const navigateHome = () => navigate(Screens.WalletHome)
+
+export const paymentRequestUnavailableScreenNavOptions = () => ({
+  ...emptyHeader,
+  headerLeft: () => <TopBarIconButton icon={<Times />} onPress={navigateHome} />,
+  headerLeftContainerStyle: { paddingLeft: 20 },
```
Comment (aaronmgdr) [inline packages/mobile/src/paymentRequest/PaymentRequestUnavailable.tsx]: move style to stylsheet?

File: packages/mobile/src/send/Send.tsx [hunk 8]
```diff
@@ -90,6 +95,42 @@ const mapDispatchToProps = {
...
+    headerLeft: () => (
+      <TopBarIconButton
+        icon={<Times />}
+        onPress={navigateBack}
+        eventName={
+          route.params?.isRequest ? CustomEventNames.send_cancel : CustomEventNames.request_cancel
+        }
+      />
+    ),
+    headerLeftContainerStyle: { paddingLeft: 20 },
```
Comment (aaronmgdr) [inline packages/mobile/src/send/Send.tsx]: style move?

File: packages/mobile/src/send/Send.tsx [hunk 9]
```diff
@@ -90,6 +95,42 @@ const mapDispatchToProps = {
...
+    headerRight: () => (
+      <TopBarIconButton
+        icon={<QRCodeBorderlessIcon height={32} color={colors.greenUI} />}
+        eventName={
+          route.params?.isRequest ? CustomEventNames.send_scan : CustomEventNames.request_scan
+        }
+        onPress={goQr}
+      />
+    ),
+    headerRightContainerStyle: { paddingRight: 16 },
```
Comment (aaronmgdr) [inline packages/mobile/src/send/Send.tsx]: style sheet?

File: packages/mobile/src/send/Send.tsx (hunk 8 above)
Comment (aaronmgdr) [reply packages/mobile/src/send/Send.tsx]: also design system doesnt use `20`

please check this shouldnt be 16 or 24

Comment (tarikbellamine) [reply packages/mobile/src/send/Send.tsx]: wasnt me, i just moved all this stuff to be consistent with your earlier suggestion. i'll change to 16

**Review by aaronmgdr — APPROVED** [review]
i only glanced.

File: packages/mobile/src/analytics/constants.ts [hunk 10]
```diff
@@ -24,42 +21,63 @@ export enum CustomEventNames {
...
+  send_secure_submit = 'send_secure_submit', // when an account is submitted for validation
+  send_secure_success = 'send_secure_success', // when an account has been validated
+  send_secure_incorrect = 'send_secure_incorrect', // when there's been an error validating the account
+  send_secure_info = 'send_secure_info', // when "help" button is pressed
+  send_secure_info_dismissed = 'send_secure_info_dismissed', // when "help" button is dismissed
+  send_secure_edit = 'send_secure_edit', // when "edit" address button is pressed to manually initate secure send flow
+
+  invite_error = 'invite_error', // when there is an error sending an invite
 
-  // Payment requests
```
Comment (annakaz) [inline packages/mobile/src/analytics/constants.ts:45]: I found these comment titles helpful, any reason to remove them?

File: packages/mobile/__mocks__/react-native-reanimated.js [hunk 11]
```diff
@@ -0,0 +1 @@
+module.exports = require('react-native-reanimated/mock')
```
Comment (annakaz) [inline packages/mobile/__mocks__/react-native-reanimated.js:1]: Does this path exist? I can't find it. Also, this seems different from the other mocks files- what is the purpose of this mock?

File: packages/react-components/analytics/CeloAnalytics.test.tsx [hunk 12]
```diff
@@ -36,17 +39,24 @@ it('tracks events with subEvents correctly', () => {
   Date.now = jest.fn(() => 1000)
   c.startTracking('mockEvent')
   Date.now = jest.fn(() => 2000)
-  c.trackSubEvent('mockEvent', 'step1')
+  c.trackSubEvent('mockEvent', 'step1', { address: '0xblablabla' })
```
Comment (annakaz) [inline packages/react-components/analytics/CeloAnalytics.test.tsx]: We often use `0xce10ce10ce10ce10ce10ce10ce10ce10ce10ce10` for arbitrary addresses

File: packages/mobile/src/analytics/constants.ts [hunk 13]
```diff
@@ -24,42 +21,63 @@ export enum CustomEventNames {
...
   outgoing_request_payment_cancel = 'outgoing_request_payment_cancel',
 
+  // Send events, separate from button tracking above
+  send_dollar_transaction = 'send_dollar_transaction',
+  send_dollar_transaction_confirmed = 'send_dollar_transaction_confirmed',
+
+  fetch_balance = 'fetch_balance',
+
   // Verification event and sub-events
+  verification = 'verification',
```
Comment (annakaz) [inline packages/mobile/src/analytics/constants.ts]: Can you remove this? No longer used after recent PR https://github.com/celo-org/celo-monorepo/pull/4050 (git merge probably missed this)

**Review by annakaz — CHANGES_REQUESTED** [review]
Great work adding all these analytics. A few minor changes!

File: packages/mobile/src/analytics/constants.ts (hunk 10 above)
Comment (tarikbellamine) [reply packages/mobile/src/analytics/constants.ts:45]: Nope it was an accident. Will add them back

File: packages/mobile/src/analytics/constants.ts (hunk 13 above)
Comment (tarikbellamine) [reply packages/mobile/src/analytics/constants.ts]: Done

File: packages/react-components/analytics/CeloAnalytics.test.tsx (hunk 12 above)
Comment (tarikbellamine) [reply packages/react-components/analytics/CeloAnalytics.test.tsx]: Gotcha, I've replaced it

File: packages/mobile/__mocks__/react-native-reanimated.js (hunk 11 above)
Comment (tarikbellamine) [reply packages/mobile/__mocks__/react-native-reanimated.js:1]: I was randomly getting animation errors on certain tests after I did a merge with master. Almost certain it was some sort of merge issue.

Honestly, I didn't dig too deep on the solution but it's how @i1skn is solving for it on [his open PR](https://github.com/celo-org/celo-monorepo/pull/4019/files#diff-162eb68a4a6cfa3ef2dd1aa2404f354f)

Comment (annakaz) [reply packages/mobile/__mocks__/react-native-reanimated.js:1]: Sounds good. Needing to add the mock in the linked PR makes more sense to me since it adds new `react-native-reanimated` imports, while this PR doesn't. I'm surprised tests weren't failing on master, but if this works, LGTM

**Review by annakaz — APPROVED** [review]

**Comment by annakaz** [comment]
Analytics logging looks great! Thanks for going through these so carefully
